### PR TITLE
common: parse time strings like "2h" or "1w2h3m23s"

### PR DESCRIPTION
**Pull request by liewegas**
- [ ] move the time parse function into ceph_time
- [ ] make 'config dump' etc show TYPE_SECS options with exact_timespan_str

### CODE REVIEW

**Comment by liewegas** [comment]
@adamemerson i'm sure that parse function could be better.. not sure where the current version came from, or if anyone really cares about parsing strings like "2 hours 5 minutes"

**Comment by tchaikov** [comment]
> not sure where the current version came from

FWIW, the current version was created by me.

File: src/common/options.cc [hunk 1]
```diff
@@ -153,30 +153,83 @@ std::chrono::seconds parse_duration(const std::string& s)
   auto secs = 0s;
   size_t start = 0;
   size_t new_start = 0;
+  using years_t = duration<int, std::ratio<365 * 3600 * 24>>;
```
Comment (tchaikov) [inline src/common/options.cc]: is there any use case of durations we should count by years and in "int"?

File: src/common/options.cc [hunk 2]
```diff
@@ -153,30 +153,83 @@ std::chrono::seconds parse_duration(const std::string& s)
...
+  using months_t = duration<int, std::ratio<30 * 3600 * 24>>;
+  using weeks_t = duration<int, std::ratio<7 * 3600 * 24>>;
```
Comment (tchaikov) [inline src/common/options.cc]: could define `weeks_t` in term of `days_t`, like
```c++
using weeks_t = duration<int, std::ratio<7 * days_t::period::num>>;
```
better readability this way.

**Comment by tchaikov** [comment]
retest this please.

**Comment by tchaikov** [comment]
retest this please.

**Comment by liewegas** [comment]
@tchaikov updated!

**Comment by tchaikov** [comment]
retest this please.

File: src/common/ceph_time.cc [hunk 3]
```diff
@@ -181,4 +182,139 @@ namespace ceph {
...
+    if (nsec) {
+      ss << ((float)nsec / 1000000000);
+    }
+    if (sec || nsec) {
+      ss << "s";
+    }
+    return ss.str();
+  }
+
+  std::chrono::seconds parse_timespan(const std::string& s)
```
Comment (jecluis) [inline src/common/ceph_time.cc]: when parsing, shouldn't we also consider ISO 8601 duration representations? e.g., P1Y1M1DT1H1M1S? I actually thought std::chrono or something similar would have a way to parse those, but I'm coming empty handed.

File: src/common/ceph_time.cc [hunk 4]
```diff
@@ -181,4 +182,139 @@ namespace ceph {
...
+    uint64_t nsec = std::chrono::nanoseconds(t).count();
+    uint64_t sec = nsec / 1000000000;
+    nsec %= 1000000000;
+    uint64_t yr = sec / (60 * 60 * 24 * 365);
+    ostringstream ss;
+    if (yr) {
+      ss << yr << "y";
+      sec -= yr * (60 * 60 * 24 * 365);
+    }
+    /* I don't like the capital M -sage
```
[the remark this answers is not in the document]
Comment (liewegas) [reply src/common/ceph_time.cc]: Yeah, I'll do 'mo'

File: src/common/ceph_time.cc (hunk 3 above)
Comment (liewegas) [reply src/common/ceph_time.cc]: (ick, why did they standardize something so ugly?)

Let's leave ISO 8601 for later?  I don't think I have encountered that duration format before, fwiw

File: src/common/ceph_time.cc (hunk 4 above)
[the remark this answers is not in the document]
Comment (tchaikov) [reply src/common/ceph_time.cc]: @liewegas do you still want to add "mo" for month?

File: src/common/ceph_time.cc [hunk 5]
```diff
@@ -181,4 +182,139 @@ namespace ceph {
...
+      while (std::isalpha(s[pos])) {
+	++pos;
+      }
+      if (unit_start != pos) {
+	string unit = s.substr(unit_start, pos - unit_start);
+	auto p = units.find(unit);
+	if (p == units.end()) {
+	  throw invalid_argument("unrecogized unit '"s + unit + "'");
+	}
+	//cout << " unit '" << unit << "' -> " << p->second << std::endl;
```
Comment (tchaikov) [inline src/common/ceph_time.cc]: might want to remove this comment.

File: src/test/common/test_config.cc [hunk 6]
```diff
@@ -159,14 +159,84 @@ TEST(md_config_t, set_val)
...
+    { "2 m5"s,  duration_cast<seconds>(seconds{2*60+5}) },
+    { "2 min5"s,  duration_cast<seconds>(seconds{2*60+5}) },
+    { "2 minutes  5"s,  duration_cast<seconds>(seconds{2*60+5}) },
+    { "1w"s, duration_cast<seconds>(seconds{3600*24*7}) },
+    { "1wk"s, duration_cast<seconds>(seconds{3600*24*7}) },
+    { "1week"s, duration_cast<seconds>(seconds{3600*24*7}) },
+    { "1weeks"s, duration_cast<seconds>(seconds{3600*24*7}) },
+    { "1month"s, duration_cast<seconds>(seconds{3600*24*30}) },
+    { "1months"s, duration_cast<seconds>(seconds{3600*24*30}) },
+    { "1M"s, duration_cast<seconds>(seconds{3600*24*30}) },
```
Comment (tchaikov) [inline src/test/common/test_config.cc]: ahh, you don't like 'M', do you? =)

File: src/common/ceph_time.cc [hunk 7]
```diff
@@ -181,4 +182,135 @@ namespace ceph {
...
+      { "hour", 60*60 },
+      { "hours", 60*60 },
+      { "d", 24*60*60 },
+      { "day", 24*60*60 },
+      { "days", 24*60*60 },
+      { "w", 7*24*60*60 },
+      { "wk", 7*24*60*60 },
+      { "week", 7*24*60*60 },
+      { "weeks", 7*24*60*60 },
+      { "mo", 30*24*60*60 },
```
Comment (tchaikov) [inline src/common/ceph_time.cc]: we either need to remove the "1M" test, or add "M" here.

**Pull request merged**